aleph-client/src/multisig.rs

```diff
@@ -0,0 +1,481 @@
+        );
+        self.finalize_xt(&connection, xt, "Cancel multisig aggregation")?;
+        Ok(())
+    }
+}
+
+type ApproveAsMultiCall = UncheckedExtrinsicV4<(
+    [u8; 2],           // call index
+    u16,               // threshold
+    Vec<AccountId32>,  // other signatories
+    Option<Timepoint>, // timepoint, `None` for initiating
+    CallHash,          // call hash
+    u64,               // max weight
+)>;
+
+type AsMultiCall = UncheckedExtrinsicV4<(
+    [u8; 2],           // call index
+    u16,               // threshold
+    Vec<AccountId32>,  // other signatories
+    Option<Timepoint>, // timepoint, `None` for initiating
+    Call,              // call
+    bool,              // whether to store call
+    u64,               // max weight
+)>;
+
+type CancelAsMultiCall = UncheckedExtrinsicV4<(
+    [u8; 2],          // call index
+    u16,              // threshold
+    Vec<AccountId32>, // other signatories
+    Timepoint,        // timepoint, `None` for initiating
+    CallHash,         // call hash
+)>;
+
+/// Implementation block containing private auxiliary methods used by those from public API.
+impl MultisigParty {
+    /// Checks whether `member_idx` is a proper position for `self.members`.
+    fn ensure_index(&self, member_idx: usize) -> Result<()> {
+        ensure!(
+            member_idx < self.members.len(),
+            MultisigError::IncorrectMemberIndex
+        );
+        Ok(())
+    }
+
+    /// For all extrinsics we have to sign it with the caller (representative) and pass
+    /// accounts of the other party members.
+    fn designate_representative_and_represented(&self, idx: usize) -> (KeyPair, Vec<AccountId>) {
+        let mut members = self.members.clone();
+        let member = members.remove(idx);
+        let others = members.iter().map(account_from_keypair).collect();
+        (member, others)
+    }
+
+    /// Reads the pallet storage and takes the timestamp regarding procedure for the `self` party
+    /// initiated at `block_hash`.
+    fn get_timestamp(
+        &self,
+        connection: &Connection,
+        call_hash: &CallHash,
+        block_hash: H256,
+    ) -> Result<Timepoint> {
+        let multisig: Multisig = connection
+            .get_storage_double_map(
+                "Multisig",
+                "Multisigs",
+                self.account.clone(),
+                *call_hash,
+                Some(block_hash),
+            )?
+            .ok_or(MultisigError::NoAggregationFound)?;
+        Ok(multisig.when)
+    }
+
+    /// Tries sending `xt` with `connection` and waits for its finalization. Returns the hash
+    /// of the containing block.
+    fn finalize_xt<T: Encode>(
+        &self,
+        connection: &Connection,
+        xt: UncheckedExtrinsicV4<T>,
+        description: &'static str,
+    ) -> Result<H256> {
+        Ok(try_send_xt(connection, xt, Some(description), Finalized)?
+            .expect("For `Finalized` status a block hash should be returned"))
```

Comment:
expect in the `try` function looks strange, maybe use here `map_error` or smth like this

Reply:
this expect is on `Option`; for `try` I've got `?`

Reply:
What I meant is that the function called `try` that panics is strange. I didn't mean that for the expect on the `Result` or `Option`

Reply:
function `try_send_xt` does not panic
function `finalize_xt` can safely unwrap option, because its security follows from inner implementation inside `try_send_xt` combined with the argument `Finalized`

Reply:
Oh, right. I was confused there.